pkg/operator/kit/watcher.go

```diff
@@ -0,0 +1,225 @@
+func NewWatcher(context KubeContext, scheme CustomResource, namespace, group, watchVersion string, callback func(event *RawEvent) error) *ResourceWatcher {
+	return &ResourceWatcher{
+		context:      context,
+		scheme:       scheme,
+		namespace:    namespace,
+		group:        group,
+		watchVersion: watchVersion,
+		callback:     callback,
+	}
+}
+
+// Watch begins watching the custom resource (TPR/CRD). The call will block until an error is raised during the watch.
+// When the watch has detected a create, update, or delete event, the raw event will be passed to the caller
+// in the callback. After the callback returns, the watch loop will continue for the next event.
+// If the callback returns an error, the error will be logged but will not abort the event loop.
+func (w *ResourceWatcher) Watch() error {
+	if w.namespace == "" {
+		logger.Infof("start watching %s resource in all namespaces at %s", w.scheme.Name, w.watchVersion)
+	} else {
+		logger.Infof("start watching %s resource in namespace %s at %s", w.scheme.Name, w.namespace, w.watchVersion)
+	}
+
+	eventCh, errCh := w.watch()
+
+	go func() {
+
+		timer := &panicTimer{
+			duration: time.Minute,
+			message:  fmt.Sprintf("unexpected long blocking (> 1 Minute) when handling %s event", w.scheme.Name),
+		}
+
+		for event := range eventCh {
+			timer.Start()
+
+			err := w.callback(event)
+			if err != nil {
+				logger.Errorf("failed to handle event %s on %s. %+v", event.Type, w.scheme.Name, err)
+			}
+
+			timer.Stop()
+		}
+	}()
+	return <-errCh
+}
+
+// watch creates a go routine, and watches the custom resource at <name>.<group> starting at
+// the given watch version. It emits events on the resources through the returned
+// event chan. Errors will be reported through the returned error chan. The go routine
+// exits on any error.
+func (w *ResourceWatcher) watch() (<-chan *RawEvent, <-chan error) {
+	eventCh := make(chan *RawEvent)
+	// On unexpected error case, the operator should exit
+	errCh := make(chan error, 1)
+
+	go func() {
+		defer close(eventCh)
+
+		for {
+			err := w.watchOuterResource(eventCh, errCh)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to watch %s resource. %+v", w.scheme.Name, err)
+				return
+			}
+		}
+	}()
+
+	return eventCh, errCh
+}
+
+func (w *ResourceWatcher) watchOuterResource(eventCh chan *RawEvent, errCh chan error) error {
+	resp, err := watchResource(w.context, w.scheme, w.namespace, w.watchVersion, w.group)
+	if err != nil {
+		errCh <- err
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("invalid status code: %s", resp.Status)
+	}
+
+	decoder := json.NewDecoder(resp.Body)
+	for {
+		ev, st, err := pollEvent(decoder)
+		done, err := handlePollEventResult(st, err, func() (bool, error) {
+			// there is no cache for the pool resource, so we return true to indicate to always reset the watch on error
```

Comment:
outdated comment since this is now a generic method? (it still refers to "pool")